### ISSUE TITLE
ClassCastException when using AuthenticatedChannel

### ISSUE THREAD

**Issue by tpj** [cryptobiu/scapi#71]
Hi,

When trying to create an authenticated channel as described in the documentation (http://scapi.readthedocs.org/en/latest/communication.html#security-of-the-connection), i.e., using this code:

	private AuthenticatedChannel createAuthenticatedChannel(Channel ch) throws Exception {
		Mac mac = new ScCbcMacPrepending(new BcAES());
		byte[] aesFixedKey = new byte[] { -61, -19, 106, -97, 106, 40, 52, -64, -115, -19, -87, -67, 98, 102, 16, 21 };
		SecretKey key = new SecretKeySpec(aesFixedKey, "AES");
		mac.setKey(key);
		return new AuthenticatedChannel(ch, mac);
	}

I get the following exception:

    java.lang.ClassCastException: edu.biu.scapi.comm.twoPartyComm.PlainTCPSocketChannel cannot be cast to edu.biu.scapi.comm.PlainChannel
    	at edu.biu.scapi.comm.ChannelDecorator.<init>(ChannelDecorator.java:46)
    	at edu.biu.scapi.comm.AuthenticatedChannel.<init>(AuthenticatedChannel.java:94)
        [snip]

At line 46 ChannelDecorator attempts to cast a PlainTCPSocketChannel to a PlainChannel, but PlainTCPSocketChannel is not an instance of PlainChannel...

**Comment by AsafKleinbort** [comment]
Hi,
Thanks for pointing it out.

As far as I see it there were two problems here.
1. ChannelDecorator assumes implicitly that it receives an instance of PlainChannel
2. PlainTCPSocketChannel is not an instance of PlainChannel

I fixed both of the issue and pushed to master in: ac7194fd5b1880ba465ddffa5b7397d571ff0d51.
Also added a (very) simple test.
If you are using scapi from source it should be enough.
If you are using a binaries you'll need to re-build. Let me know if you need any help.

BTW, with the current version you can workaround it by using another channel, e.g., PlainTCPChannel.

**Comment by tpj** [comment]
Hi,

I still have a little trouble with this one. I've now updated my SCAPI git repo to latest version (14570f9991de496d89054ac965c53738f4689b46). I would like to create a "bleeding edge" jar of that version (since the correction of this issue came after 2.4.0). But running "make compile-scapi" gives a lot of compile errors (I managed to fix some of them by corrcting the ant script, but not all..). Running just "make clean" followed by "make" and "make install" produces a jar Scapi-2.4.0.jar in /usr/lib/scapi, but that seems to actually be 2.4.0. At least, using that jar, I still get the above ClassCastException.

So my question is: How can I obtain a scapi jar containing the most recent commits, in particular the https://github.com/cryptobiu/scapi/commit/ac7194fd5b1880ba465ddffa5b7397d571ff0d51

Also, I wanted to try using PlainTCPChannel instead of PlainTCPSocketChannel, to do a work-around with the current (2.3.0) jar. But I couldn't find a way to instantiate PlainTCPChannel.

Thanks in advance.

**Comment by AsafKleinbort** [comment]
Hi,

You are right. Currently running make will only build the underlying libs and the JNI, and "make compile-scapi" was broken.

I fixed "compile scapi" and pushed it to master (dfc8b2da384d87310c8f755b301bc085b1557671). Notice that there is an additional prerequisite now - need to install junit4 (e.g., `sudo apt-get install junit4` on ubuntu systems) - I also updated the docs.

The ant script assumes that the junit jar reside in `/usr/share/java`. If it is anywhere else on you system just fix this in the build.xml file.

Also notice that the jar created reside in `~/scapi/bulid/scapi` so if you are using `scapic` and `scapi` to run and compile your code you'll have to replace the jar in assets folders with the jar that you built.

Hope it will work for you now :)
Let me know.
